enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
+
+#### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+- Document SLOs for the component
+- Conduct load testing
+- User facing documentation created in
+  [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+#### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+This new behavior will be added in 4.12 as part of the installation
+configurations for customers to utilize.
+
+Enabling the feature after installation for HA/3NC is not supported in 4.12, so
+we do not need to address what happens if an older cluster upgrades and then the
+feature is turned on.
+
+When upgrades occur for current single node deployments we will need to set the
+global identifier during the upgrade. We will do this via the NTO and the
+trigger for this event will be:
+
+- If the `capacity` field set on the master node and is running in Single Node
+
+We will not change the current machine configs for single node deployments if
+they are already set, this will be done to avoid extra restarts. We will need to
+be clear with customers however, if they add the
+`spec.workloads[Infrastructure]` we will then generate the new machine config
+and an extra restart will happen. They will need to delete the old machine
+configs afterwards.
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+The addition to the API is an optional field which should not require any
+conversion admission webhook changes. This change will only be used to allow the
+user to explicitly define their intent and simplify the machine manifest by
+generating the extra machine manifests that are currently being created
+independently of the `PerformanceProfile` CRD.
+
+Futhermore the design and scope of this enhancement will mean that the existing
+Admission webhook will continue to apply the same warnings and error messages to
+Pods as described in the [failure modes](#failure-modes).
+
+#### Failure Modes
+
+In a failure situation, we want to try to keep the cluster operational.
+Therefore, there are a few conditions under which the admission hook will strip
+the workload annotations and add an annotation `workload.openshift.io/warning`
+with a message warning the user that their partitioning instructions were
+ignored. These conditions are:
+
+1. When a pod has the Guaranteed QoS class
+2. When mutation would change the QoS class for the pod
+3. When the feature is inactive because not all nodes are reporting the
+   management resource
+
+#### Support Procedures
+
+N/A
+
+## Implementation History
+
+WIP
+
+## Alternatives
+
+N/A
```

Comment:
We discussed not adding the field to the install-config. It would be good to capture the reason we decided we needed it after all.

We could do nothing and say this feature is for SNO deployments only.

We could not support SNO upgrades and simplify the implementation. Being explicit about why we feel we must support that upgrade path, beyond saying that customers want it, would be good.

Reply:
Alternative I would like to discuss:

All workload partitioning trigger logic will be installer based (or based on a pre-install CR). PerfProfile will have no extra API at all, but NTO will update the necessary cri-o files when WP is detected as enabled in the cluster transparently behind the scenes.

This will keep clean API that stays cluster wide for WP and MCP wide for PerfProfile with no way for the user to try to reconfigure or remove WP from an installed cluster.

We can also revisit the PAO render mode integration with the installer now that PAO logic is part of the core payload. This would allow generating all the MC, Tuned and KubeletConfig manifests at day 0.